File: firstMissingPositive.js
/**
 * @param {number[]} A
 * @return {number}
 */
var firstMissingPositive = function(A) {
    var len = A.length;
    var i = 0;
    var j = 0;
    while(i < len){
        if(A[i] <= len && A[i] > 0 && A[i] != A[A[i]-1]){
            var tmp = A[A[i]-1];
            A[A[i]-1] = A[i];
            A[i] = tmp;
        } else{
            i++;
        }
    }
    
    while(j < len){
        if(A[j] != j+1) return j+1;
        j++;
    }
    return len+1;
    
};

File: firstMissingPositive.java
public class Solution {
    public int firstMissingPositive(int[] A) {
        int n = A.length;
        for(int i = 0; i< n; i++){
            if(A[i] <= n && A[i] > 0 && A[i] != A[A[i] -1]){
                int temp = A[A[i]-1];  
                A[A[i]-1] = A[i];  
                A[i] = temp; 
                i--;
            }
        }
        int j = 0;
        for(int i = 0; i < n; i++){
            if(A[i] != (i+1)) return i+1;
        }
        return n+1;
    }
}